phantazm-mob/src/main/java/com/github/phantazmnetwork/mob/config/MobModelConfigProcessor.java

```diff
@@ -0,0 +1,131 @@
+                throw new ConfigProcessException("goal creators are not a list");
+            }
+
+            ConfigList goalCreatorsList = goalCreatorsElement.asList();
+            Collection<GoalCreator> goalCreators = new ArrayList<>(goalCreatorsList.size());
+            for (ConfigElement goalCreatorElement : goalCreatorsList) {
+                goalCreators.add(goalCreatorConfigProcessor.dataFromElement(goalCreatorElement));
+            }
+
+            goalCreatorsGroups.add(goalCreators);
+        }
+
+        String displayNameString = element.getStringOrDefault((String) null, "displayName");
+        Component displayName = displayNameString == null ? null : miniMessage.deserialize(displayNameString);
+
+        ConfigNode equipmentNode = element.getNodeOrThrow("equipment");
+        Map<EquipmentSlot, ItemStack> equipment = new HashMap<>(equipmentNode.size());
+        for (Map.Entry<String, ConfigElement> entry : equipmentNode.entrySet()) {
+            EquipmentSlot equipmentSlot = EQUIPMENT_SLOT_MAP.get(entry.getKey().toUpperCase());
+            if (equipmentSlot == null) {
+                throw new ConfigProcessException("unknown equipment slot");
+            }
+
+            if (!entry.getValue().isString()) {
+                throw new ConfigProcessException("equipment value is not string");
+            }
+
+            equipment.put(equipmentSlot, itemStackConfigProcessor.dataFromElement(entry.getValue()));
+        }
+
+        float maxHealth = element.getNumberOrThrow("maxHealth").floatValue();
+
+        return new MobModel(descriptor, goalCreatorsGroups, displayName, equipment, maxHealth);
+    }
+
+    @Override
+    public @NotNull ConfigElement elementFromData(@NotNull MobModel model) throws ConfigProcessException {
+        ConfigElement descriptor = descriptorProcessor.elementFromData(model.getDescriptor());
+
+        ConfigList goalCreatorsGroups = new ArrayConfigList();
+        for (Iterable<GoalCreator> creators : model.getGoalCreatorsGroups()) {
+            ConfigList goalCreatorsList = new ArrayConfigList();
+            for (GoalCreator creator : creators) {
+                goalCreatorsList.add(goalCreatorConfigProcessor.elementFromData(creator));
+            }
+
+            goalCreatorsGroups.add(goalCreatorsList);
+        }
+
+        ConfigElement displayName = model.getDisplayName()
+                .map(miniMessage::serialize)
+                .map(ConfigPrimitive::new)
+                .orElseGet(() -> new ConfigPrimitive(null));
+
+        ConfigNode equipmentNode = new LinkedConfigNode();
+        for (Map.Entry<EquipmentSlot, ItemStack> entry : model.getEquipment().entrySet()) {
+            equipmentNode.put(entry.getKey().name().toUpperCase(), itemStackConfigProcessor.elementFromData(entry.getValue()));
+        }
+
+        ConfigElement maxHealth = new ConfigPrimitive(model.getMaxHealth());
+
+        ConfigNode element = new LinkedConfigNode();
```

Comment:
(will stop commenting on LinkedCfgNodes w/o explicit capacity but i think all of them should have it)